### DISCUSSION TITLE
`props` follows the one-way data flow strictly

### DISCUSSION THREAD

**Discussion by edison1105** [vuejs/rfcs#279] [General]
`props` follows the one-way data flow principle, i.e. it is not allowed to reassign values to props in subcomponents. However, if `props` is passed as an object, it is still possible to modify the value of the object's properties in the child component. If we strictly follow the single data flow, we should not even modify object properties. `Vue2` is limited by `defineProperty` which may not be good for this. But `Vue3` is very easy to implement. 
It may be a `break change`, but it will avoid a lot of bugs.
For example:
```javascript
    props: {
        direction: {
            type:String,
           default:'down',
            validator (val) {
                val = 'up' // this will change direction value in DEV,but not work in PROD
                return true
            }
        },
       someObject:{
           type:Object
       }
    },
    created(){
        // change the props value and parent component doesn't know this.
        this.someObject.foo = 100
    }
```
The `val` can be modified in the development environment, but in the production environment the `val` will not be modified and there is no hint. difference behavier in development and production(Vue2 and Vue3 both have this problem)

if `props` follows the one-way data flow strictly.we can show some warnings when developer try to modify `props` in DEV mode.I think this will help developer avoid bugs.

**Comment by posva** [comment]
This could be useful as a plugin but in practice, a lot of people use this (native js) behaviour so I don't think it should exist out of the box

**Reply by edison1105** [reply]
Maybe add a configuration that exposes it for use by `vite` or `webpack`. When the user turns on `{ readonlyProps: true }`, modifying the props gives a warning in DEV environment .
It can effectively prevent data confusion and ensure that only the parent component can modify the data.
So, I think it might be a good start for users to slowly follow a real one-way data flow.

**Reply by posva** [reply]
It could be definitely interesting as an opt-in 👍

**Comment by jods4** [comment]
There are very valid cases for modifying an object passed as a prop.

Consider a form component to edit a person:
```vue
<template>
  <form>
    <label>
      First name:
      <input v-model="person.firstName" />
    </label>
  </form>
</template>

<script>
  export default {
    props: { person: Object }
  }
</script>
```

This component mutates its `person` props and that's something many people would rely on, I think.

**Comment by stafyniaksacha** [comment]
Why `props` are not wrapped in [`readonly`](https://v3.vuejs.org/guide/reactivity-fundamentals.html#prevent-mutating-reactive-objects-with-readonly) function ?

**Reply by jods4** [reply]
It would create tons of proxy-related problems?

**Reply by edison1105** [reply]
Yes, props wrapped in `readonly` will generate a lot of problems.
But we can do it in `PublicInstanceProxyHandlers` to make more sense.
For example:
```javascript
 // in get()
case AccessTypes.PROPS:
   const val = props![key]
   return val ==='object'? shallowReadonly(val): val
```
Only effective in DEV.
